components/Banner/examples/supplemental/MDCBannerView.m

```diff
@@ -54,124 +67,527 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
-        MDCButton *button = [[MDCButton alloc] init];
-        [mutableButtons addObject:button];
-        [self addSubview:button];
-      }
-    } else if (numberOfButtons < self.buttons.count) {
-      for (NSUInteger index = numberOfButtons; index < self.buttons.count; ++index) {
-        MDCButton *lastObject = [mutableButtons lastObject];
-        [lastObject removeFromSuperview];
-        [mutableButtons removeLastObject];
-      }
-    }
-    self.buttons = [mutableButtons copy];
+- (UIImage *)icon {
+  return self.iconImageView.image;
+}
+
+- (void)setButtons:(NSMutableArray<MDCButton *> *_Nonnull)buttons {
+  NSAssert(buttons.count <= kNumberOfButtonsLimit, @"%@ doesn't support more than %lu buttons",
+           NSStringFromClass([self class]), (unsigned long)kNumberOfButtonsLimit);
+  _buttons = buttons;
+  for (MDCButton *button in buttons) {
+    button.translatesAutoresizingMaskIntoConstraints = NO;
+    [self.containerView addSubview:button];
   }
+  self.isLayoutDirty = YES;
+}
+
+- (void)setBackgroundColor:(UIColor *)backgroundColor {
+  [super setBackgroundColor:backgroundColor];
+  self.containerView.backgroundColor = backgroundColor;
+}
+
+- (void)setTextFont:(UIFont *)textFont {
+  self.textLabel.font = textFont;
+}
+
+- (UIFont *)textFont {
+  return self.textLabel.font;
+}
+
+- (void)setTextColor:(UIColor *)textColor {
+  self.textLabel.textColor = textColor;
+}
+
+- (UIColor *)textColor {
+  return self.textLabel.textColor;
 }
 
-- (NSUInteger)numberOfButtons {
-  return self.buttons.count;
+- (void)setIconTintColor:(UIColor *)iconTintColor {
+  self.iconImageView.tintColor = iconTintColor;
+}
+
+- (UIColor *)iconTintColor {
+  return self.iconImageView.tintColor;
 }
 
 #pragma mark - UIView overrides
 
 - (CGSize)sizeThatFits:(CGSize)size {
-  self.layout = [[MDCBannerViewLayout alloc] initWithSizeToFit:size
-                                                     textLabel:self.textLabel
-                                                imageContainer:self.iconImageViewContainer
-                                                       buttons:self.buttons];
-  return [self.layout frameSize];
+  return CGSizeMake([UIScreen mainScreen].bounds.size.width, self.layout.size.height);
 }
 
-- (void)layoutSubviews {
-  [super layoutSubviews];
+- (CGSize)intrinsicContentSize {
+  return CGSizeMake([UIScreen mainScreen].bounds.size.width, self.layout.size.height);
+}
 
-  [self applyMDCBannerViewLayout:self.layout];
+- (void)updateConstraints {
+  if (!self.cachedLayout || self.cachedLayout != self.layout) {
+    [self updateConstraintsWithLayoutStyle:self.layout.style];
+    self.cachedLayout = self.layout;
+  }
+
+  [super updateConstraints];
 }
 
 #pragma mark - Layout methods
 
-- (void)applyMDCBannerViewLayout:(MDCBannerViewLayout *)layout {
-  if (![self isMDCBannerViewLayoutApplicable:layout]) {
-    return;
+- (void)setIsLayoutDirty:(BOOL)isLayoutDirty {
```

Comment:
Is this intended to be used similarly to `-setNeedsLayout`? Am I understanding that right?

Reply:
My `intrinsicContentSize` is calculated based on `layout`. My original intention was use this method to make sure `layout` is recalculated and `intrinsicContentSize` can use the latest value. It would be good if there is a better/cleaner to do achieve that. Any suggestion?

Reply:
I'm not sure. With the text field stuff I've been working on I've basically just been calculating everything on every UIKit layout pass, and I don't have my own method of marking things as dirty outside of `-setNeedsLayout`. I do see the appeal of it though. Haven't thought through the problems one might run into yet.

Speaking of `intrinsicContentSize`, I do agree with the stuff Rob was saying [here](https://github.com/material-components/material-components-ios/pull/6830#discussion_r264399411).

Reply:
Tip: 

I agree with @andrewoverton that optimizing for "dirty" layouts should only happen if there's a documented need (e.g., performance measurements showing layout is too expensive). Methods like `layoutSubviews` or constraints updates happen only when the view hierarchy changes in a meaningful way - it shouldn't happen frequently enough on a Banner to justify caching layouts.

Please **strongly consider** performing layout on every pass. If you use layout constraints, just keep them up to date unless you find them slow enough to justify `updateConstraints`.

Reply:
Thanks for the comments @andrewoverton and @romoore. I think it resolved some worries I had on making this part of logic not clear. I will change it accordingly.

@romoore For the **strongly consider** part, do you recommend to remove the `updateConstraints` and directly call a method to update constraint when text font is changed (for example). Or can I keep the current way that `setNeedUpdateConstraints` when text font is changed?

Reply:
Removed `isLayoutDirty` in cd2ed349f6c8d0710689b1f5672d37f5d7f81434.

Thanks for the suggestions.